processor/src/fast/checkpoints.rs

```diff
@@ -0,0 +1,647 @@
+    /// the pre-recorded address
+    pub fn replay_block_address(&mut self) -> Felt {
+        self.block_addresses.pop_front().expect("No block address operations recorded")
+    }
+
+    /// Replays a `Hasher::permute` operation, returning its address and result
+    pub fn replay_permute(&mut self) -> (Felt, [Felt; 12]) {
+        self.permutation_operations
+            .pop_front()
+            .expect("No permutation operations recorded")
+    }
+
+    /// Replays a Merkle path verification, returning the pre-recorded address and computed root
+    pub fn replay_build_merkle_root(&mut self) -> (Felt, Word) {
+        self.build_merkle_root_operations
+            .pop_front()
+            .expect("No build merkle root operations recorded")
+    }
+
+    /// Replays a Merkle root update, returning the pre-recorded address, old root, and new root
+    pub fn replay_update_merkle_root(&mut self) -> (Felt, Word, Word) {
+        self.mrupdate_operations.pop_front().expect("No mrupdate operations recorded")
+    }
+}
+
+/// Implements a shim for stack overflow operations, in which all overflow values and addresses
+/// during a given fragment are pre-recorded by the fast processor and replayed by the main trace
+/// fragment generators.
+///
+/// This is used to simulate stack overflow functionality in parallel trace generation without
+/// needing to maintain the actual overflow table. All overflow operations are recorded during
+/// fast execution and then replayed during parallel trace generation.
+///
+/// The shim records overflow values (from pop operations) and overflow addresses (representing
+/// the clock cycle of the last overflow update) and provides replay methods that return the
+/// pre-recorded values. This works naturally since the fast processor has exactly the same
+/// access patterns as the main trace generators.
+#[derive(Debug)]
+pub struct StackOverflowReplay {
+    /// Recorded overflow values and overflow addresses from pop_overflow operations. Each entry
+    /// represents a value that was popped from the overflow stack, and the overflow address of the
+    /// at the top of the overflow stack *after* the pop operation.
+    overflow_values: VecDeque<(Felt, Felt)>,
+
+    /// Recorded (stack depth, overflow address) returned when restoring a context
+    restore_context_info: VecDeque<(usize, Felt)>,
+}
+
+impl Default for StackOverflowReplay {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl StackOverflowReplay {
+    /// Creates a new StackOverflowReplay with empty operation vectors
+    pub fn new() -> Self {
+        Self {
+            overflow_values: VecDeque::new(),
+            restore_context_info: VecDeque::new(),
+        }
+    }
+
+    // MUTATORS
+    // --------------------------------------------------------------------------------
+
+    /// Records the value returned by a pop_overflow operation, along as the overflow address stored
```

Comment:
"... along as..." is it intentional, maybe can be re-written to simplify the phrasing